libhb/enc_qsv.c

```diff
@@ -1342,54 +1413,75 @@ int encqsvInit(hb_work_object_t *w, hb_job_t *job)
-        if (pv->param.videoParam->mfx.RateControlMethod == MFX_RATECONTROL_CQP)
+        // set the GOP structure for x264
+        if (pv->param.gop.gop_ref_dist < 0)
         {
-            // ensure B-pyramid is enabled for CQP on Haswell
-            pv->param.gop.gop_pic_size = 32;
+            if (pv->param.videoParam->mfx.RateControlMethod == MFX_RATECONTROL_CQP)
+            {
+                pv->param.gop.gop_ref_dist = 4;
+            }
+            else
+            {
+                pv->param.gop.gop_ref_dist = 3;
+            }
         }
-        else
+
+        // set the keyframe interval
+        if (pv->param.gop.gop_pic_size < 0)
         {
-            // set the keyframe interval based on the framerate
-            pv->param.gop.gop_pic_size = rate;
+            int rate = (double)job->orig_vrate.num / job->orig_vrate.den + 0.5;
+            if (pv->param.videoParam->mfx.RateControlMethod == MFX_RATECONTROL_CQP)
+            {
+                // ensure B-pyramid is enabled for CQP on Haswell
+                pv->param.gop.gop_pic_size = 32;
+            }
+            else
+            {
+                // set the keyframe interval based on the framerate
+                pv->param.gop.gop_pic_size = rate;
+            }
         }
     }
+    pv->param.videoParam->mfx.GopRefDist = pv->param.gop.gop_ref_dist;
     pv->param.videoParam->mfx.GopPicSize = pv->param.gop.gop_pic_size;
 
     // sanitize some settings that affect memory consumption
-    if (!pv->is_sys_mem)
+    if (pv->param.videoParam->mfx.CodecId == MFX_CODEC_HEVC)
     {
-        // limit these to avoid running out of resources (causes hang)
-        pv->param.videoParam->mfx.GopRefDist   = FFMIN(pv->param.videoParam->mfx.GopRefDist,
-                                                       pv->param.rc.lookahead ? 8 : 16);
-        pv->param.codingOption2.LookAheadDepth = FFMIN(pv->param.codingOption2.LookAheadDepth,
-                                                       pv->param.rc.lookahead ? (48 - pv->param.videoParam->mfx.GopRefDist -
-                                                                                 3 * !pv->param.videoParam->mfx.GopRefDist) : 0);
+        // HEVC encoding takes a lot of memory .. you got to live with it.
+        pv->param.videoParam->mfx.GopRefDist   = FFMIN(pv->param.videoParam->mfx.GopRefDist, 256);
     }
     else
     {
-        // encode-only is a bit less sensitive to memory issues
-        pv->param.videoParam->mfx.GopRefDist   = FFMIN(pv->param.videoParam->mfx.GopRefDist, 16);
-        pv->param.codingOption2.LookAheadDepth = FFMIN(pv->param.codingOption2.LookAheadDepth,
-                                                       pv->param.rc.lookahead ? 100 : 0);
+        if (!pv->is_sys_mem)
+        {
+            // limit these to avoid running out of resources (causes hang)
+            pv->param.videoParam->mfx.GopRefDist   = FFMIN(pv->param.videoParam->mfx.GopRefDist,
```

Comment:
GopRefDist is distance between I- or P (or GPB) - key frames based on https://software.intel.com/en-us/node/628473
I do not see the case where we need to have so large number.

Reply:
Closing this conversation as other than the below "of interest" discussion .. the problems with >64 surfaces in d3dll textures and lookahead support are included in other conversations above. The following is mainly of interest and shouldn't be considered important reading.

The large number for GopRefDist (256) is the specification defined maximum allowed value for hevc and is only really useful for the highly specific case of hierarchical B frames in HEVC. The huge improvements for inter-frame matching for HEVC allow long tracks and 256 frames is about 10 seconds at 30fps. It isn't uncommon for video to have little movement for large parts of the screen or some object of central focus for 10 seconds (e.g. a zoom on talking people) . Because information that is visible for short periods is much less visually important than information visible for longer periods the B frame uses less bitrate on short lived information. This is quite intuitive as you can't read many letters if they are visible for <5-10 frames .. but can be understood less intuitively as the relationship between measurement, uncertainty, time, space and scale. For the handbrake case .. higher frequency information in time and space is generally (a) less visible and (b) can have a disproportionate effect on QP despite not being easily observed.

To get a feel .. experiment with powers of 2 in gop-ref-dist in the below cmd line (adjust the q value so you're starting to get some visual artifacts):

`HandBrakeCLI -i TourDeFrance.mpv  -o test.mkv  -e qsv_h265 --encoder-preset quality -q 26 -E copy -x "force-cqp=1:gop-ref-dist=32:gop-pic-size=512"  --disable-qsv-decoding
`

and you'll hopefully see  improvement in file sizes coming at little visual cost during playback (it may be more visible if paused). To get the biggest effect use simple animation inputs (e.g. some modern and all older cartoons or anime) and expect less effect in video with constant scene changes and high frequency data (higher resolution computer games recordings or some music videos spring to mind). I've even seen improvements of order 30-40x on 4k video for hierarchical QP HEVC vs ICQ HEVc.